test/powershell/Language/Scripting/CommandErrorHandling.Tests.ps1

```diff
@@ -0,0 +1,470 @@
+        $PSCmdlet.ThrowTerminatingError($er)
+
+        Write-Verbose -Verbose "verbose-message"
+    }
+
+    function ErrorActionStop {
+        [CmdletBinding()]
+        param()
+
+        Get-Command NonExist -ErrorAction Stop
+        Write-Verbose -Verbose "verbose-message"
+    }
+
+    function ThrowException {
+        [CmdletBinding()]
+        param()
+
+        throw 'throw-exception'
+        Write-Verbose -Verbose "verbose-message"
+    }
+
+    function WriteErrorAPI {
+        [CmdletBinding()]
+        param()
+
+        $ex = [System.ArgumentException]::new('arg-exception')
+        $er = [System.Management.Automation.ErrorRecord]::new($ex, 'WriteErrorAPI:error', 'InvalidArgument', $null)
+        $PSCmdlet.WriteError($er)
+
+        Write-Verbose -Verbose "verbose-message"
+    }
+
+    function WriteErrorCmdlet {
+        [CmdletBinding()]
+        param()
+
+        Write-Error 'write-error-cmdlet'
+        Write-Verbose -Verbose "verbose-message"
+    }
+
+    function MethodInvocationThrowException {
+        [CmdletBinding()]
+        param()
+
+        ## This method call throws exception.
+        $iss = [initialsessionstate]::Create()
+        $iss.ImportPSModule($null)
+
+        Write-Verbose -Verbose "verbose-message"
+    }
+
+    function ExpressionThrowException {
+        [CmdletBinding()]
+        param()
+
+        1/0 ## throw exception.
+        Write-Verbose -Verbose "verbose-message"
+    }
+
+'@).Invoke()
+
+        function RunCommand {
+            param(
+                [string] $Command,
+                [ValidateSet('Continue', 'Ignore', 'SilentlyContinue', 'Stop')]
+                [string] $ErrorAction
+            )
+
+            $pwsh.Commands.Clear()
+            $pwsh.Streams.ClearStreams()
+            $pwsh.AddCommand($command).AddParameter('ErrorAction', $ErrorAction).Invoke()
+        }
+
+        function RunScript {
+            param([string] $Script)
+
+            $pwsh.Commands.Clear()
+            $pwsh.Streams.ClearStreams()
+            $pwsh.AddScript($Script).Invoke()
+        }
+
+        function GetLastError {
+            $pwsh.Commands.Clear()
+            $pwsh.AddCommand('Get-Error').Invoke()
+        }
+
+        function ClearDollarError {
+            $pwsh.Commands.Clear()
+            $pwsh.AddScript('$Error.Clear()').Invoke()
+        }
+    }
+
+    AfterAll {
+        $pwsh.Dispose()
+    }
+
+    Context 'Terminating error' {
+
+        It "'ThrowTerminatingError' stops execution with 'ActionPreference.<ErrorAction>' for error action" -TestCases @(
+            @{ ErrorAction = 'Continue' }
+            @{ ErrorAction = 'Ignore' }
+            @{ ErrorAction = 'SilentlyContinue' }
+        ) {
+            param ($ErrorAction)
+
+            $failure = $null
+
+            try {
+                RunCommand -Command 'ThrowTerminatingError' -ErrorAction $ErrorAction
```

Comment:
`ThrowTerminatingError`  implementation (and others test functions too) ignore `ErrorAction` parameter. Why do we need to run the test (and follows) for some test cases?

Also why do we skip Break and Inquire values.

Reply:
> `ThrowTerminatingError` implementation (and others test functions too) ignore `ErrorAction` parameter.

This test and the the one below it are to verify that behavior: the terminating error always stops execution without being affected by the specified error actions.

`break` breaks into debugger (even for terminating errors) and `Inquire` would also trigger interactive prompt for non-terminating errors, so both are not covered in the new tests. But I did manually test `break` and `inquire` for the `clean` block.

One more clarification: the tests in `CommandErrorHandling.Tests.ps1` are all for verifying the existing error handling behaviors in command execution. The purpose is to make sure adding the `clean` block doesn't cause a regression in the existing error handling.

Reply:
> > `ThrowTerminatingError` implementation (and others test functions too) ignore `ErrorAction` parameter.
> 
> This test and the the one below it are to verify that behavior: the terminating error always stops execution without being affected by the specified error actions.

I don't understand why we add ErrorAction parameter to RunCommand but ignore it in test function like ThrowTerminatingError. If we want really test that behavior ThrowTerminatingError should  use ErrorAction parameter too.

> One more clarification: the tests in CommandErrorHandling.Tests.ps1 are all for verifying the existing error handling behaviors in command execution. The purpose is to make sure adding the clean block doesn't cause a regression in the existing error handling.

I'd suggest to merge them in separate PR before this PR.

Reply:
I understand why you think that may be better, but I don't think it's necessary.

Reply:
The PR is too large and this could save a lot of time on the code review.

Reply:
Will move the regression protection tests to a separate PR.